File: src/emuchef_editor/api/session.py
"""In-process document sessions for the editor JSON API."""

from __future__ import annotations

from dataclasses import dataclass
from uuid import uuid4

from emuchef_editor.core.documents.recipe_document import RecipeDocument
from emuchef_editor.core.yaml.loader import create_recipe_document_from_template, load_recipe_document

from .command_codec import decode_recipe_command
from .dto import command_result_to_dto, diagnostic_to_dto, document_to_dto, ref_index_to_dto, step_specs_to_dto
from .errors import ApiError
from .protocol import failure, success


@dataclass(slots=True)
class DocumentSession:
    document_id: str
    document: RecipeDocument


class DocumentSessionManager:
    """Owns live editor documents for future sidecar-style API calls."""

    def __init__(self) -> None:
        self._sessions: dict[str, DocumentSession] = {}

    def open_recipe(self, path: str, authored_root: str | None = None) -> dict:
        try:
            document = load_recipe_document(path, authored_root=authored_root)
        except Exception as exc:
            return failure(ApiError("load_failed", f"Failed to load recipe: {exc}", {"path": path}), exc=exc)
        return self._add_document(document)

    def create_recipe_from_template(
        self,
        template_path: str,
        destination_path: str,
        recipe_id: str,
        authored_root: str | None = None,
    ) -> dict:
        try:
            document = create_recipe_document_from_template(
                template_path,
                destination_path=destination_path,
                recipe_id=recipe_id,
                authored_root=authored_root,
            )
        except Exception as exc:
            return failure(
                ApiError(
                    "load_failed",
                    f"Failed to create recipe from template: {exc}",
                    {"templatePath": template_path, "destinationPath": destination_path},
                ),
                exc=exc,
            )
        return self._add_document(document)

    def close_document(self, document_id: str) -> dict:
        session = self._sessions.pop(document_id, None)
        if session is None:
            return self._unknown_document(document_id)
        return success({})

    def get_document(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        return success({"document": document_to_dto(session.document, document_id=session.document_id)})

    def save_recipe(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        try:
            session.document.save()
        except Exception as exc:
            return failure(ApiError("save_failed", f"Failed to save recipe: {exc}", {"documentId": document_id}), exc=exc)
        return success({"document": document_to_dto(session.document, document_id=session.document_id)})

    def save_recipe_as(self, document_id: str, path: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        try:
            session.document.save_as(path)
        except Exception as exc:
            return failure(
                ApiError("save_failed", f"Failed to save recipe as: {exc}", {"documentId": document_id, "path": path}),
                exc=exc,
            )
        return success({"document": document_to_dto(session.document, document_id=session.document_id)})

    def emit_yaml(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        return success({"yaml": session.document.to_yaml()})

    def apply_recipe_command(self, document_id: str, command: dict) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        try:
            decoded_command = decode_recipe_command(command)
        except ApiError as exc:
            return failure(exc, exc=exc)
        try:
            changed = session.document.apply_command(decoded_command)
        except Exception as exc:
            return failure(ApiError("command_failed", f"Command failed: {exc}", {"documentId": document_id}), exc=exc)
        return success(
            {
                "commandResult": command_result_to_dto(changed=changed),
                "document": document_to_dto(session.document, document_id=session.document_id),
            }
        )

    def undo(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        changed = session.document.undo()
        return success(
            {
                "commandResult": command_result_to_dto(changed=changed),
                "document": document_to_dto(session.document, document_id=session.document_id),
            }
        )

    def redo(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        changed = session.document.redo()
        return success(
            {
                "commandResult": command_result_to_dto(changed=changed),
                "document": document_to_dto(session.document, document_id=session.document_id),
            }
        )

    def validate(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        try:
            result = session.document.validate()
        except Exception as exc:
            return failure(ApiError("validation_failed", f"Validation failed: {exc}", {"documentId": document_id}), exc=exc)
        return success({"diagnostics": [diagnostic_to_dto(diagnostic) for diagnostic in result.diagnostics]})

    def get_ref_index(self, document_id: str) -> dict:
        session = self._session(document_id)
        if session is None:
            return self._unknown_document(document_id)
        return success({"refIndex": ref_index_to_dto(session.document.ref_index)})

    def list_step_specs(self) -> dict:
        return success({"stepSpecs": step_specs_to_dto()})

    def _add_document(self, document: RecipeDocument) -> dict:
        document_id = str(uuid4())
        self._sessions[document_id] = DocumentSession(document_id=document_id, document=document)
        return success({"document": document_to_dto(document, document_id=document_id)})

    def _session(self, document_id: str) -> DocumentSession | None:
        return self._sessions.get(document_id)

    @staticmethod
    def _unknown_document(document_id: str) -> dict:
        return failure(
            ApiError(
                "unknown_document",
                f"Unknown document id: {document_id}",
                {"documentId": document_id},
            )
        )

File: crates/emuchef-rust-backend/src/session.rs
//! Sidecar-only document session manager for the experimental Rust backend.
//!
//! Session state is intentionally in-memory and process-local. It persists
//! across JSONL requests handled by one sidecar process and is never shared with
//! one-shot CLI invocations.

use indexmap::IndexMap;
use serde_json::{json, Value};

use crate::commands;
use crate::document::RecipeDocument;
use crate::dto;
use crate::errors::ApiError;
use crate::ref_index;
use crate::user_configuration_document::{self, UserConfigurationDocument};

#[derive(Debug, Default)]
pub struct DocumentSessionManager {
    documents: IndexMap<String, RecipeDocument>,
    user_configurations: IndexMap<String, UserConfigurationDocument>,
    next_id: u64,
}

impl DocumentSessionManager {
    pub fn open_recipe(
        &mut self,
        path: &str,
        authored_root: Option<&str>,
    ) -> Result<Value, ApiError> {
        let document = RecipeDocument::open(path, authored_root).map_err(|error| {
            ApiError::load_failed(
                format!("Failed to load recipe: {error}"),
                json!({ "path": path }),
            )
        })?;
        let document_id = self.next_document_id();
        let document_dto = dto::document_to_dto(&document, &document_id);
        self.documents.insert(document_id, document);
        Ok(json!({ "document": document_dto }))
    }

    pub fn create_recipe_from_template(
        &mut self,
        template_path: &str,
        destination_path: &str,
        recipe_id: &str,
        authored_root: Option<&str>,
    ) -> Result<Value, ApiError> {
        let document = RecipeDocument::create_from_template(
            template_path,
            destination_path,
            recipe_id,
            authored_root,
        )
        .map_err(|error| {
            ApiError::load_failed(
                format!("Failed to create recipe from template: {error}"),
                json!({ "templatePath": template_path, "destinationPath": destination_path }),
            )
        })?;
        let document_id = self.next_document_id();
        let document_dto = dto::document_to_dto(&document, &document_id);
        self.documents.insert(document_id, document);
        Ok(json!({ "document": document_dto }))
    }

    pub fn get_document(&self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document(document_id)?;
        Ok(json!({
            "document": dto::document_to_dto(document, document_id)
        }))
    }

    pub fn save_recipe(&mut self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        document.save().map_err(|error| {
            ApiError::save_failed(
                format!("Failed to save recipe: {error}"),
                json!({ "documentId": document_id }),
            )
        })?;
        Ok(json!({
            "document": dto::document_to_dto(document, document_id)
        }))
    }

    pub fn save_recipe_as(&mut self, document_id: &str, path: &str) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        document.save_as(path).map_err(|error| {
            ApiError::save_failed(
                format!("Failed to save recipe as: {error}"),
                json!({ "documentId": document_id, "path": path }),
            )
        })?;
        Ok(json!({
            "document": dto::document_to_dto(document, document_id)
        }))
    }

    pub fn apply_recipe_command(
        &mut self,
        document_id: &str,
        command: &Value,
    ) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        let command = commands::decode_recipe_command(command)?;
        let changed = document.apply_command(command).map_err(|error| {
            ApiError::command_failed(
                format!("Command failed: {error}"),
                json!({ "documentId": document_id }),
            )
        })?;
        Ok(json!({
            "commandResult": {"changed": changed},
            "document": dto::document_to_dto(document, document_id),
        }))
    }

    pub fn undo(&mut self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        let changed = document.undo();
        Ok(json!({
            "commandResult": {"changed": changed},
            "document": dto::document_to_dto(document, document_id),
        }))
    }

    pub fn redo(&mut self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        let changed = document.redo();
        Ok(json!({
            "commandResult": {"changed": changed},
            "document": dto::document_to_dto(document, document_id),
        }))
    }

    pub fn emit_yaml(&self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document(document_id)?;
        Ok(json!({ "yaml": document.yaml() }))
    }

    pub fn validate(&mut self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        document.validate();
        Ok(json!({ "diagnostics": document.diagnostics() }))
    }

    pub fn get_ref_index(&self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.document(document_id)?;
        Ok(json!({
            "refIndex": ref_index::ref_index_to_dto(document.recipe())
        }))
    }

    pub fn set_document_authored_root(
        &mut self,
        document_id: &str,
        authored_root: Option<&str>,
    ) -> Result<Value, ApiError> {
        let document = self.document_mut(document_id)?;
        document.set_authored_root(authored_root);
        Ok(json!({
            "document": dto::document_to_dto(document, document_id)
        }))
    }

    pub fn close_document(&mut self, document_id: &str) -> Result<Value, ApiError> {
        if self.documents.shift_remove(document_id).is_none() {
            return Err(ApiError::unknown_document(document_id));
        }
        Ok(json!({}))
    }

    pub fn open_user_configuration(
        &mut self,
        path: &str,
        authored_root: Option<&str>,
    ) -> Result<Value, ApiError> {
        let document = UserConfigurationDocument::open(path, authored_root).map_err(|error| {
            user_configuration_load_error("load", error, json!({ "path": path }))
        })?;
        let document_id = self.next_document_id();
        let document_dto = user_configuration_document::document_to_dto(&document, &document_id);
        self.user_configurations.insert(document_id, document);
        Ok(json!({ "document": document_dto }))
    }

    #[allow(clippy::too_many_arguments)]
    pub fn create_user_configuration(
        &mut self,
        path: &str,
        configuration_id: &str,
        name: &str,
        device_plan: &str,
        selected_recipes: Vec<String>,
        authored_root: Option<&str>,
    ) -> Result<Value, ApiError> {
        let document = UserConfigurationDocument::create(
            path,
            configuration_id,
            name,
            device_plan,
            selected_recipes,
            authored_root,
        )
        .map_err(|error| user_configuration_load_error("create", error, json!({ "path": path })))?;
        let document_id = self.next_document_id();
        let document_dto = user_configuration_document::document_to_dto(&document, &document_id);
        self.user_configurations.insert(document_id, document);
        Ok(json!({ "document": document_dto }))
    }

    pub fn get_user_configuration_document(&self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.user_configuration_document(document_id)?;
        Ok(json!({
            "document": user_configuration_document::document_to_dto(document, document_id)
        }))
    }

    pub fn save_user_configuration(&mut self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        document.save().map_err(|error| {
            ApiError::save_failed(
                format!("Failed to save user configuration: {error}"),
                json!({ "documentId": document_id }),
            )
        })?;
        Ok(json!({
            "document": user_configuration_document::document_to_dto(document, document_id)
        }))
    }

    pub fn save_user_configuration_as(
        &mut self,
        document_id: &str,
        path: &str,
    ) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        document.save_as(path).map_err(|error| {
            ApiError::save_failed(
                format!("Failed to save user configuration as: {error}"),
                json!({ "documentId": document_id, "path": path }),
            )
        })?;
        Ok(json!({
            "document": user_configuration_document::document_to_dto(document, document_id)
        }))
    }

    pub fn set_user_configuration_binding(
        &mut self,
        document_id: &str,
        key: &str,
        value: Option<Value>,
    ) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        let changed = document.set_binding(key, value).map_err(|error| {
            ApiError::command_failed(
                format!("User-configuration binding command failed: {error}"),
                json!({ "documentId": document_id, "key": key }),
            )
        })?;
        Ok(json!({
            "commandResult": { "changed": changed },
            "document": user_configuration_document::document_to_dto(document, document_id),
        }))
    }

    pub fn set_user_configuration_selected_recipes(
        &mut self,
        document_id: &str,
        selected_recipes: Vec<String>,
    ) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        let changed = document
            .set_selected_recipes(selected_recipes)
            .map_err(|error| {
                ApiError::command_failed(
                    format!("User-configuration recipe command failed: {error}"),
                    json!({ "documentId": document_id }),
                )
            })?;
        Ok(json!({
            "commandResult": { "changed": changed },
            "document": user_configuration_document::document_to_dto(document, document_id),
        }))
    }

    pub fn set_user_configuration_device_plan(
        &mut self,
        document_id: &str,
        device_plan: String,
    ) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        let changed = document.set_device_plan(device_plan).map_err(|error| {
            ApiError::command_failed(
                format!("User-configuration device-plan command failed: {error}"),
                json!({ "documentId": document_id }),
            )
        })?;
        Ok(json!({
            "commandResult": { "changed": changed },
            "document": user_configuration_document::document_to_dto(document, document_id),
        }))
    }

    pub fn validate_user_configuration(&mut self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        document.validate();
        Ok(json!({ "diagnostics": document.diagnostics() }))
    }

    pub fn emit_user_configuration_yaml(&self, document_id: &str) -> Result<Value, ApiError> {
        let document = self.user_configuration_document(document_id)?;
        Ok(json!({ "yaml": document.yaml() }))
    }

    pub fn set_user_configuration_authored_root(
        &mut self,
        document_id: &str,
        authored_root: Option<&str>,
    ) -> Result<Value, ApiError> {
        let document = self.user_configuration_document_mut(document_id)?;
        document.set_authored_root(authored_root);
        Ok(json!({
            "document": user_configuration_document::document_to_dto(document, document_id)
        }))
    }

    pub fn close_user_configuration(&mut self, document_id: &str) -> Result<Value, ApiError> {
        if self.user_configurations.shift_remove(document_id).is_none() {
            return Err(ApiError::unknown_document(document_id));
        }
        Ok(json!({}))
    }

    fn next_document_id(&mut self) -> String {
        self.next_id += 1;
        format!("doc-{}", self.next_id)
    }

    fn document(&self, document_id: &str) -> Result<&RecipeDocument, ApiError> {
        self.documents
            .get(document_id)
            .ok_or_else(|| ApiError::unknown_document(document_id))
    }

    fn document_mut(&mut self, document_id: &str) -> Result<&mut RecipeDocument, ApiError> {
        self.documents
            .get_mut(document_id)
            .ok_or_else(|| ApiError::unknown_document(document_id))
    }

    fn user_configuration_document(
        &self,
        document_id: &str,
    ) -> Result<&UserConfigurationDocument, ApiError> {
        self.user_configurations
            .get(document_id)
            .ok_or_else(|| ApiError::unknown_document(document_id))
    }

    fn user_configuration_document_mut(
        &mut self,
        document_id: &str,
    ) -> Result<&mut UserConfigurationDocument, ApiError> {
        self.user_configurations
            .get_mut(document_id)
            .ok_or_else(|| ApiError::unknown_document(document_id))
    }
}

fn user_configuration_load_error(
    action: &str,
    error: crate::user_configuration::UserConfigurationLoadError,
    details: Value,
) -> ApiError {
    ApiError::load_failed(
        format!("Failed to {action} user configuration: {error}"),
        details,
    )
}
